internal/controller/driver_controller.go

```diff
@@ -18,45 +18,1033 @@ package controller
 
 import (
 	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"maps"
+	"reflect"
+	"regexp"
+	"slices"
+	"strings"
 
+	"github.com/go-logr/logr"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	storagev1 "k8s.io/api/storage/v1"
+	k8sErrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/utils/ptr"
 	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
 	"sigs.k8s.io/controller-runtime/pkg/client"
-	"sigs.k8s.io/controller-runtime/pkg/log"
+	ctrlutil "sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	ctrllog "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
 
-	csiv1alpha1 "github.com/ceph/ceph-csi-operator/api/v1alpha1"
+	csiv1a1 "github.com/ceph/ceph-csi-operator/api/v1alpha1"
+	"github.com/ceph/ceph-csi-operator/internal/utils"
 )
 
+//+kubebuilder:rbac:groups=csi.ceph.io,resources=drivers,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=csi.ceph.io,resources=drivers/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=csi.ceph.io,resources=drivers/finalizers,verbs=update
+//+kubebuilder:rbac:groups=csi.ceph.io,resources=operatorconfigs,verbs=get;list;watch
+
+type DriverType string
+
+const (
+	RbdDriverType    = "rbd"
+	CephFsDriverType = "cephfs"
+	NfsDriverType    = "nfs"
+)
+
+const ownerRefAnnotationKey = "csi.ceph.io/ownerref"
+
+// A regexp used to parse driver short name and driver type from the
+// driver's full name
+var nameRegExp, _ = regexp.Compile(fmt.Sprintf(
+	`^(.*)\.(%s|%s|%s)\.csi\.ceph\.com$`,
+	RbdDriverType,
+	CephFsDriverType,
+	NfsDriverType,
+))
```

Comment:
Why write manual parsing code where regexp is a specialized tool to deal with these use cases? 
The regexp here is very simple and translates to anything that exactly ends with one of the following:
`.rbd.csi.ceph.com`, `.cephfs.csi.ceph.com`, or `.nfs.csi.ceph.com`

Your example above will not match anything because we do not allow the string "driver', but an analogous example that would parse is `thing1.thing2.driver.csi.ceph.com.thing3.rbd.csi.ceph.com` which will correctly be splitter into [`thing1.thing2.driver.csi.ceph.com.thing3`, `rbd`, `csi.ceph.com`]

Reply:
by `driver` I meant one of the drivers (rbd/cephfs/nfs). Now the regex became even more complicated (relative to my earlier understanding) after refactor. I only meant a simple parsing would be reasonable rather than using a regex which could potentially need a test.

now I see `(.+)` (earlier `(.*)`) both of which allows anything, even spaces probably which anyways isn't allowed by kube api but the regex is more forgiving than an higher api. Again I'm not picking, to me it just feels regex isn't required.

can be resolved as I already made my point 😃.

Reply:
> you because of a lack of proficiency in "RexExp lang"
- a very bad take 😓, I said `a little rusty` not lacking.

> The regexp is limiting on top of what is already limited by the API itself.
- the regexp here allows what is already limited by API, that's what I meant.

Nevertheless, could be resolved.

Reply:
> a very bad take 😓, I said a little rusty not lacking.
@leelavg  I am sorry, I did not mean to offend, just to state a point of view regarding any reader's ability, not you in specific. 

> the regexp here allows what is already limited by API, that's what I meant.
I am failing to see that. The regexp here guarantees additional constraints on top of what the API allows and helps extract that information. So it achieves both at the same time.

Reply:
resolve.